File: CountLines/CountLines3.c
#include <stdio.h>
#include <string.h>

// http://stackoverflow.com/a/4278920/5086987

#define BUFFER_SIZE 8192

int countlines( char *filename ) {
  int lines = 0;

  FILE *file = fopen( filename, "r" );
  if ( file != NULL ) {
    char  buffer[ BUFFER_SIZE + 1 ];
    char* pointer;
    int   bytes;
    while ( ! feof( file ) ) {
      bytes = fread( buffer, sizeof( char ), BUFFER_SIZE, file );
      if ( bytes <= 0 ) {
        break;
      }

      buffer[ bytes ] = '\0';
      for ( pointer = buffer; pointer; pointer = strchr( pointer, '\n' ) ) {
        lines++;
        ++pointer;
      }
    }

    fclose( file );
  }

  return lines;
}

int main( int argc, char *argv[] ) {
  printf( "%d\n", countlines( argv[ 1 ] ) );
}

File: CountLines/CountLines3.go
package main

import (
  "fmt"
  "os"
  "io"
  "bufio"
  "bytes"
)

const
  BUFFER_SIZE = 8 * 1024

// http://stackoverflow.com/a/24563853/5086987
func CountLines( filename string ) int {
  lines := 0

  hasCharacters := false

  file, err := os.Open( filename )
  if err == nil {
    newLine := []byte{'\n'}
    buffer  := make( []byte, BUFFER_SIZE )
    reader  := bufio.NewReader( file )

    defer file.Close()

    for {
      size, err := reader.Read( buffer )
      if err != nil && err != io.EOF {
        return 0
      }

      lines += bytes.Count( buffer[ : size ], newLine )

      if err == io.EOF {
        break
      }

      hasCharacters = buffer[ size - 1 ] != '\n';
    }

    if hasCharacters {
      lines += 1
    }
  }

  return lines
}

func main() {
  fmt.Println( CountLines( os.Args[ 1 ] ) )
}
